### PR TITLE
refactoring graph to reuse similar code

### PR DESCRIPTION
**Pull request by baykaner**
sharing implementation between graph AddTrainables and graph InsertSharedCopy

### CODE REVIEW

File: libs/math/include/math/base_types.hpp [hunk 1]
```diff
@@ -113,5 +114,53 @@ fetch::meta::IfIsFloat<T, T> static function_tolerance()
   return T(1e-6);
 }
 
+template <typename T>
+static constexpr meta::IfIsNonFixedPointArithmetic<T, bool> is_nan()
+{
+  return std::fetestexcept(FE_INVALID);
+}
```
Comment (evsmithx) [inline libs/math/include/math/base_types.hpp:148]: std::isnan(foo) is a function for checking if a particular value is NaN, and this function checks the state, so I think the name could be clearer?

File: libs/ml/include/ml/utilities/min_max_scaler.hpp [hunk 2]
```diff
@@ -55,47 +54,31 @@ template <typename TensorType>
...
-    }
+  x_range_ = x_max_ - x_min_;
 
-    ++x_min_it;
-    ++x_max_it;
-    ++x_range_it;
-  }
+  std::cout << "x_range_: " << x_range_ << std::endl;
+  std::cout << "x_max_: " << x_max_ << std::endl;
+  std::cout << "x_min_: " << x_min_ << std::endl;
```
Comment (evsmithx) [inline libs/ml/include/ml/utilities/min_max_scaler.hpp]: Leftover debugging code?

File: libs/ml/include/ml/utilities/min_max_scaler.hpp [hunk 3]
```diff
@@ -145,44 +123,63 @@ void MinMaxScaler<TensorType>::DeNormalise(TensorType const &input_tensor,
...
+}
+
+}  // namespace utilities
+}  // namespace ml
+
+namespace serializers {
+
 /**
- * Initialise min, max, and range
+ * serializer for OpLayerNormSaveableParams saveable params
```
Comment (evsmithx) [inline libs/ml/include/ml/utilities/min_max_scaler.hpp]: Wrong docstring

File: scripts/etch-language-tests/ml/scaler.etch [hunk 4]
```diff
@@ -10,6 +10,7 @@ function main()
     data_tensor.fillRandom();
 
     var scaler = Scaler();
+    scaler.setScale(data_tensor.At(0, 0), data_tensor.At(0, 0));
```
Comment (evsmithx) [inline scripts/etch-language-tests/ml/scaler.etch]: I think lines 13 and 14 do the same thing?

File: libs/vm_modules/src/ml/utilities/scaler.cpp [hunk 5]
```diff
@@ -79,11 +86,27 @@ void VMScaler::Bind(Module &module)
...
 
+bool VMScaler::SerializeTo(serializers::MsgPackSerializer &buffer)
+{
+  buffer << *this;
+  return true;
+}
+
+bool VMScaler::DeserializeFrom(serializers::MsgPackSerializer &buffer)
+{
+  buffer.seek(0);
```
Comment (evsmithx) [inline libs/vm_modules/src/ml/utilities/scaler.cpp]: Is seek(0) correct here? There might have been previous objects serialised in the buffer.

File: libs/vm_modules/include/vm_modules/ml/utilities/scaler.hpp [hunk 6]
```diff
@@ -63,10 +67,48 @@ class VMScaler : public fetch::vm::Object
...
 };
 
 }  // namespace utilities
 }  // namespace ml
 }  // namespace vm_modules
+
+namespace serializers {
+
+/**
+ * serializer for VMOptimiser
```
Comment (evsmithx) [inline libs/vm_modules/include/vm_modules/ml/utilities/scaler.hpp]: VMScaler not VMOptimiser

**Review by evsmithx — CHANGES_REQUESTED** [review]

**Review by evsmithx — DISMISSED** [review]
LGTM

File: libs/ml/tests/ml/ops/log.cpp [hunk 7]
```diff
@@ -131,9 +96,14 @@ TYPED_TEST(LogBothTest, backward_test)
 
   ASSERT_TRUE(prediction.at(0).AllClose(gt, fetch::math::function_tolerance<DataType>(),
                                         fetch::math::function_tolerance<DataType>()));
+  fetch::math::state_clear<DataType>();
 }
 
-TYPED_TEST(LogBothTest, saveparams_test)
+//// TODO(1195): fixed point and floating point tests should be unified.
+//  using DataType = typename TypeParam::Type;
+//  assert(fetch::math::state_division_by_zero<DataType>() == false);
```
Comment (MissingNO57) [inline libs/ml/tests/ml/ops/log.cpp]: Leftover comments?

**Review by MissingNO57 — CHANGES_REQUESTED** [review]

**Review by MissingNO57 — APPROVED** [review]
LGTM

**Pull request merged**